server/proxy/proxy.go

```diff
@@ -178,3 +187,67 @@ func (p *Proxy) upstreamForApp(app *apps.App) (upstream.Upstream, error) {
 		return nil, utils.NewInvalidError("not a valid app type: %s", app.AppType)
 	}
 }
+
+func (p *Proxy) CleanUserCallContext(userID string, cc *apps.Context) (*apps.Context, error) {
+	ctx := &apps.Context{
+		ContextFromUserAgent: cc.ContextFromUserAgent,
+	}
+
+	var postID, channelID, teamID string
+
+	switch {
+	case ctx.PostID != "":
+		postID = ctx.PostID
+
+		post, err := p.mm.Post.GetPost(postID)
+		if err != nil {
+			return nil, err
+		}
+
+		channelID = post.ChannelId
+
+		_, err = p.mm.Channel.GetMember(channelID, userID)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to get channel membership. user=%v channel=%v", userID, channelID)
+		}
+
+		c, err := p.mm.Channel.Get(channelID)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to get channel. channel=%v", channelID)
+		}
+
+		teamID = c.TeamId
+
+	case ctx.ChannelID != "":
+		channelID = ctx.ChannelID
+
+		_, err := p.mm.Channel.GetMember(ctx.ChannelID, userID)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to get channel membership. user=%v channel=%v", userID, ctx.ChannelID)
+		}
+
+		c, err := p.mm.Channel.Get(channelID)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to get channel. channel=%v", channelID)
+		}
+
+		teamID = c.TeamId
+
+	case ctx.TeamID != "":
+		teamID = ctx.TeamID
+
+		_, err := p.mm.Team.GetMember(teamID, userID)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to get team membership. user=%v team=%v", userID, teamID)
+		}
+
+	default:
+		return nil, errors.Errorf("no post, channel, or team context provided. user=%v", userID)
```

Comment:
0/5 on how to handle this case. We don't currently have any use cases where we are submitting a call outside of the context of a given team, so I figured it should be required to provide at least a post id, channel id, or team id in the request.